reducer/countdistinct.go

```diff
@@ -0,0 +1,43 @@
+package reducer
+
+import (
+	"github.com/axiomhq/hyperloglog"
+	"github.com/mccanne/zq/pkg/zeek"
+	"github.com/mccanne/zq/pkg/zson"
+)
+
+// CountDistinct uses hyperloglog to approximate the count of unique values for
+// a field.
+type CountDistinct struct {
+	Reducer
+	Field  string
+	sketch *hyperloglog.Sketch
+}
+
+func NewCountDistinct(name, field string) *CountDistinct {
+	return &CountDistinct{
+		Reducer: New(name),
+		Field:   field,
+		sketch:  hyperloglog.New(),
+	}
+}
+
+func (c *CountDistinct) Consume(r *zson.Record) {
+	i, ok := r.Descriptor.LUT[c.Field]
+	if !ok {
+		return
+	}
+	v := r.Slice(i)
+	c.sketch.Insert(v)
+}
+
+func (c *CountDistinct) Result() zeek.Value {
+	return &zeek.Count{Native: c.sketch.Estimate()}
+}
+
+// Sketch returns the native structure used to compute the distinct count
```

Comment:
I'm not sure how this can be currently used in a flowgraph but we can figure it out when we need it.

Reply:
Yeah, I was thinking of the case where I'd want to merge split countdistinct flowgraphs into a single flowgraph. I wanted the ability to access the underlying data structure when I get to this. I was thinking I might just wrap this in shell proc and make a custom "hyperloglog" type that implements the zeek.Value interface. Of course the downstream merging countdistinct would need to know what to do with this special value.